framework/validators/IpValidator.php

```diff
@@ -0,0 +1,498 @@
+ *   false  - negation is not allowed
+ *
+ * @author SilverFire <d.naumenko.a@gmail.com>
+ * @since 2.0.7
+ */
+class IpValidator extends Validator
+{
+    /**
+     * The length of IPv6 address in bits
+     */
+    const IPV6_ADDRESS_LENGTH = 128;
+
+    /**
+     * The length of IPv4 address in bits
+     */
+    const IPV4_ADDRESS_LENGTH = 32;
+
+    /**
+     * Default negation char when [[negationChar]] is set to `true`
+     * @see negationChar
+     */
+    const DEFAULT_NEGATION_CHAR = '!';
+
+    /**
+     * @var boolean whether support of IPv6 addresses is enabled
+     */
+    public $ipv6 = true;
+
+    /**
+     * @var boolean whether support of IPv4 addresses is enabled
+     */
+    public $ipv4 = true;
+
+    /**
+     * @var boolean whether the address can be a CIDR subnet
+     *    true - the subnet is required
+     *   false - the address can not have the subnet
+     *    null - ths subnet is optional
+     */
+    public $subnet = false;
+
+    /**
+     * @var boolean whether to add the prefix with the smallest length (32 for IPv4 and 128 for IPv6)
+     * to an address without it.
+     * Works only when attribute [[subnet]] is not false.
+     * Example: `10.0.1.5` will normalized to `10.0.1.5/32`, `2008:db0::1` will be normalized to `2008:db0::1/128`
+     * @see subnet
+     */
+    public $normalize = false;
+
+    /**
+     * @var string|boolean
+     * @see negationChar
+     */
+    public $_negationChar = false;
+
+    /**
+     * @var boolean whether to expand an IPv6 address to the full notation format
+     */
+    public $expandIPv6 = false;
+
+    /**
+     * @var array IPv4 or IPv6 ranges that are allowed or denied.
+     *
+     * The array should contain only `allow` or/and  `deny` keys.
+     * Their order determines the order of checks of the IP address.
+     *
+     * When the first is `allow` - all the `allow` IP addresses will be checked, and at least one must
+     * match or the validation will fail. Then all the `deny` (if present) will be checked, if one of them
+     * matched the validation will fail.
+     * At last, if the IP address is neither on the `allow`` nor on the `deny`, the validation will also fail.
+     *
+     * When the first is `deny` - all the `deny` IP addresses will be checked, if one matched the validation
+     * will fail, unless it is also present on the `allow`.
+     * If it is not found on any of the lists the validation will be passed.
+     *
+     * Tip: it is useful to set `deny` first, when it is necessary to deny a less specific subnet and
+     * to allow a more specific one. The example below will result in passing `192.168.1.1`,
+     * but `192.168.2.1` will be denied:
+     *
+     * ```php
+     * 'ips' => [
+     *     'deny' => ['192.168.0.0/16'],
+     *     'allow' => ['192.168.1.0/24']
+     * ]
+     * ```
+     *
+     * @see isAllowed()
+     */
+    public $ips = [];
```

Comment:
@samdark I like your idea, we will get an [nginx-styled](http://nginx.org/en/docs/http/ngx_http_access_module.html?&_ga=1.102502791.783966358.1445265302#allow) config.

What do you think about adding the magic:

```php
public $networks = [
    '*' => ['0.0.0.0/0', '::/0'],
    'all' => ['0.0.0.0/0', '::/0'],
    'private' => ['10.0.0.0/8', '172.16.0.0/12', '192.168.0.0/16', 'fd00::/8'],
    'multicast' => ['224.0.0.0/24', '224.0.1.0', '224.0.1.0/24', '224.0.2.0/23', '224.0.4.0/22', '224.0.8.0/21', '224.0.16.0/20', '224.0.32.0/19', '224.0.64.0/18', '224.0.128.0/17', '224.1.0.0/16', '224.2.0.0/15', '224.4.0.0/14', '224.8.0.0/13', '224.16.0.0/12', '224.32.0.0/11', '224.64.0.0/10', '224.128.0.0/9', '225.0.0.0/8', '226.0.0.0/7', '228.0.0.0/6', '232.0.0.0/6', '236.0.0.0/7', '238.0.0.0/8', '239.0.0.0/8', 'ff00::/8'],
];
```

```php
[
  '!192.168.0.0/24', // deny 
  'all' // allow all
]
```


```php
[
  '192.168.0.0/24', // allow one rage of private networks
  '!private', // deny other private networks
  'all'  // allow all
]
```